File: library/src/test/java/com/example/practice/java/modern/tech/sandbox/library/hello/lambda/ImplementDataStructureByLambda.java
package com.example.practice.java.modern.tech.sandbox.library.hello.lambda;

import java.util.function.Function;

public class ImplementDataStructureByLambda {
    public static void main(String[] args) {
        final Function<Boolean, Integer> intPair = createPair(1, 2);
        System.out.printf("Pair: %s %s%n", first(intPair), second(intPair));

        final Function<Boolean, String> stringPair = createPair("hello", "world");
        System.out.printf("Pair: %s %s%n", first(stringPair), second(stringPair));

        final Function<Boolean, Object> pair = createPair(1, "hello");
        System.out.printf("Pair: %s %s%n", first(pair), second(pair));
    }

    private static <T> Function<Boolean, T> createPair(T first, T second) {
        return m -> {
            if (m) return first;
            else return second;
        };
    }

    private static <T> T first(Function<Boolean, T> pair) {
        return pair.apply(true);
    }


    private static <T> T second(Function<Boolean, T> pair) {
        return pair.apply(false);
    }
}

File: library/src/test/java/com/example/practice/java/modern/tech/sandbox/library/hello/lambda/ImplementDataStructureByLambda.kt
package com.example.practice.java.modern.tech.sandbox.library.hello.lambda

fun main(args: Array<String>) {
    val intPair = createPair(1, 2)
    println("Pair: ${first(intPair)} ${second(intPair)}")

    val stringPair = createPair("hello", "world")
    println("Pair: ${first(stringPair)} ${second(stringPair)}")

    val pair = createPair(1, "hello")
    println("Pair: ${first(pair)} ${second(pair)}")
}

fun <T> createPair(first: T, second: T): Pair<T> = { selector: Boolean ->
    if (selector) first else second
}

fun <T> first(pair: Pair<T>) = pair(true)
fun <T> second(pair: Pair<T>) = pair(false)

typealias Pair<T> = (Boolean) -> T
